File: src/parser.cpp
#include "lexer.cpp"

#include <vector>
#include <string>

namespace ct {
    using str = std::string;

    template<typename T>
    using vec = std::vector<T>;

    template<typename... T>
    using var = std::variant<T...>;

    struct Node {};

    struct Type : Node {};
    struct Builtin : Type {
        enum { 
            U8, U16, U32, U64,
            I8, I16, I32, I64
        } type;
    };

    struct Decl : Node {};

    struct Alias : Decl {
        Alias(str n, Type* t)
            : name(n)
            , type(t)
        { }
        str name;
        Type* type;
    };

    struct Unit : Node {
        vec<Decl*> decls;
    };

    struct Parser {
        Parser(Lexer l)
            : ahead(Token::invalid)
            , lex(l)
        { }

        Type* parse_type() {
            
        }

        Alias* parse_alias() {
            auto name = expect<Token::ident>();
            expect<Token::key>(Keyword::ASSIGN);
            auto type = parse_type();
            expect<Token::key>(Keyword::SEMICOLON);
            return new Alias(name, type);
        }

        Decl* parse_decl() {
            if (consume<Token::key>(Keyword::ALIAS)) {
                return parse_alias();
            }

            return nullptr;
        }

        Unit parse() {

        }

        template<Token::type_t T, typename V>
        bool consume(const V& val) {
            auto tok = next();
            if (tok.is(T, val)) {
                return true;
            }
            ahead = tok;
            return false;
        }

        template<Token::type_t T>
        Token::underlying<T>::type expect() {
            auto tok = next();
            if (tok.is(T)) {
                return std::get<typename Token::underlying<T>::type>(tok.data);
            }
            printf("bbbb\n");
            std::abort();
        }

        template<Token::type_t T, typename V>
        void expect(const V& val) {
            if (expect<T>() != val) {
                printf("Aaaaaa\n");
                std::abort();
            }
        }

        Token next() {
            auto tok = ahead;
            if (tok.type == Token::invalid) {
                tok = lex.next();
            }
            
            ahead = Token(Token::invalid);
            
            return tok;
        }

        Token ahead;
        Lexer lex;
    };
}

File: src2/Parser.kt

// expressions

open class Node()

open class Stmt()
open class Expr : Stmt()

class StmtList(val stmts: List<Stmt>) : Stmt()

// types

open class Type : Node()

typealias Field = Pair<String, Type>

class Struct(
    val fields: List<Field>
) : Type()

class Tuple(val fields: List<Type>) : Type()

class Func(
    val args: List<Field>,
    val ret: Type?,
    val body: Stmt
) : Node()

// toplevel stuff

class Import(
    val path: List<String>,
    val items: List<String>
)

typealias TypeMap = Map<String, Type>
typealias FuncMap = Map<String, Func>

class Unit(
    val imports: List<Import>,
    val types: TypeMap,
    val funcs: FuncMap
)



class Parser(var lex: Lexer) {
    private var ahead: Token? = null

    private fun next(): Token {
        val tok = ahead ?: lex.next()
        ahead = null
        return tok
    }


    private fun ident(): String {
        return (next() as Ident).id
    }

    private fun consume(key: Keyword): Boolean {
        val tok = next()
        if (tok is Key && tok.key == key) {
            return true
        }
        ahead = tok
        return false
    }

    private fun collect(sep: Keyword): List<String> {
        val items = mutableListOf<String>()

        do items.add(ident()) while (consume(sep))

        return items.toList()
    }

    private fun expect(key: Keyword) {
        val tok = next()
        assert(tok is Key && tok.key == key)
    }

    private fun parseName(): List<String> {
        return collect(Keyword.COLON2)
    }

    private fun parseImportItems(): List<String> {
        return collect(Keyword.COMMA)
    }

    private fun parseImport(): Import {
        val path = parseName()

        expect(Keyword.LPAREN)

        val items = parseImportItems()

        expect(Keyword.RPAREN)

        return Import(path, items)
    }

    private fun parseImports(): List<Import> {
        val imports = mutableListOf<Import>()

        while (consume(Keyword.IMPORT)) {
            imports.add(parseImport())
            expect(Keyword.SEMICOLON)
        }

        return imports
    }

    private fun parseTypeName(): Type {

    }

    private fun parseTuple(): Tuple {
        val fields = mutableListOf<Type>()

        do fields.add(parseType()) while (consume(Keyword.COMMA))

        expect(Keyword.RPAREN)

        return Tuple(fields)
    }

    private fun parseType(): Type {
        val type: Type

        while (true) {
            if (consume(Keyword.MUL)) {
                type =
            }
        }
    }

    private fun parseFuncArgs(): List<Field> {
        if (!consume(Keyword.LPAREN)) {
            return listOf()
        }


    }

    private fun parseFuncReturn(): Type? {
        if (!consume(Keyword.COLON)) {
            return null
        }

        return parseType()
    }

    private fun parseExpr(): Expr {

    }

    private fun parseStmtList(): StmtList {

    }

    private fun parseFuncBody(): Stmt {
        if (consume(Keyword.ASSIGN)) {
            return parseExpr()
        } else {
            return parseStmtList()
        }
    }

    private fun parseFunc(): Pair<String, Func> {
        val name = ident()
        val args = parseFuncArgs()
        val ret = parseFuncReturn()
        val body = parseFuncBody()

        return Pair(name, Func(args, ret, body))
    }

    private fun parseStruct(): Pair<String, Struct> {

    }

    private fun parseAttribute() {

    }

    private fun parseBody(): Pair<TypeMap, FuncMap> {
        val types = mutableMapOf<String, Type>()
        val funcs = mutableMapOf<String, Func>()


        while (true) {
            if (consume(Keyword.DEF)) {
                val (name, func) = parseFunc()
                funcs.put(name, func)
            } else if (consume(Keyword.STRUCT)) {
                val (name, type) = parseStruct()
                types.put(name, type)
            } else if (consume(Keyword.AT)) {
                parseAttribute()
            } else {
                break
            }
        }

        return Pair(types, funcs)
    }

    fun parse(): Unit {
        val imports = parseImports()

        val (types, funcs) = parseBody()

        return Unit(imports, types, funcs)
    }
}